### DISCUSSION TITLE
inconsistent search results for IP addresses

### DISCUSSION THREAD

**Discussion by tfboy** [netbox-community/netbox#9490] [Q&A]
I seem to be getting inconsistent results when doing partial IP address searches.
If I have for example, I have an IP defined as 192.168.5.1/24 and search for "5.1", it won't find it. But it will list other IPs that have a 5.1 in it (10.5.10.1, 10.5.12.23, etc).
If I search for the complete IP address, with or without the subnet, it finds it.
I haven't figured out if it's related to a VRF being defined or not. It seems to sometimes not search within VRFs and only report results in the Global VRF, but then it does list the 10.5.10.1, 10.5.12.23 which are in a specific, non-Global VRF.

I've tried a few variations hoping to figure out a pattern which could explain why, but no luck so far.

I appear to get the same strange results whether I do a global search or an IP-address search.

What am I missing?

netbox 3.2.4

**Comment by candlerb** [comment]
Where are you searching?
1. in the global search box with "All Objects" selected
2. in the global search box with "IP addresses" selected
3. in the 'quick search' box at the top of `/ipam/ip-addresses/`
4. in the Search box under the Filters tab at `/ipam/ip-addresses/` - I think this is the same as (2)
5. somewhere else?

In my testing, it matches string prefixes only - e.g. "10.12.254.1" is found if I search for "10.1", but not "12.2".

If "5.1" is finding 10.5.10.1, it could be for some other reason - e.g. the string 5.1 appears within the Description or DNS Name for that IP address (in which case a substring match is permitted). See the search logic [here](https://github.com/netbox-community/netbox/blob/v3.2.4/netbox/ipam/filtersets.py#L563-L571).  As far as I can see, the search does not care about VRF.  However, this might vary depending on exactly which search entry point you're using.

FYI there is a related but different issue for prefixes - #6016.

**Comment by tfboy** [comment]
Searching in the global search box with all objects (default).
If I put for example just "5.1", I get a bunch of prefixes and the following for IP addresses:
![image](https://user-images.githubusercontent.com/6972726/172390298-e1db41af-1d60-42d4-94a3-07a1fa1273a3.png)

If I put in the whole IP, I get whole IPs and IPs that have that element of the string in:
![image](https://user-images.githubusercontent.com/6972726/172390886-8bf6f5fa-a39f-458e-bcbe-01212fabc5d3.png)

Now I can understand why in the second case, it doesn't give the results from the first search. But I cannot see why the first search gives just the results from the first screenshot and doesn't include the IPs of the second. The only explanation I can think of is that the partial IP address search only looks in the Global VRF for IPs as the ones in the first screenshot are in Global whereas the ones in the second are in a defined VRF. But, as you say, I wouldn't have expected it to care about VRFs.

The IPs and even the "5.1" bit does not appear in any description or DNS. It's only referenced as an IP address.

**Reply by kkthxbye-code** [reply]
The search query is simple:

```
qs_filter = (
    Q(dns_name__icontains=value) |
    Q(description__icontains=value) |
    Q(address__istartswith=value)
)
```

The query has to either be a substring of the dns name/description or the ip address must start with it. I tried adding your example IP addresses and searched for 5.1 and none of them showed up.

> The IPs and even the "5.1" bit does not appear in any description or DNS. It's only referenced as an IP address.

Could you please double check as what you are suggesting shouldn't be possible.

**Reply by candlerb** [reply]
What Netbox version? What Postgres version?

Does it make a difference if you select "IP addresses" from the dropdown next to the global search box?

I can't reproduce what you see here (with Netbox 3.2.4 and Postgres 12.11).  On your system, probably the most direct way to find out what's going on is to turn on SQL query logging to see what actual database search queries are being issued. 

It may be possible to turn on query logging either in Postgres itself, or at the Django end.  For the latter, I tried this in `configuration.py`:

```
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'file': {
            'level': 'DEBUG',
            'class': 'logging.FileHandler',
            'filename': '/var/tmp/netbox.log',
        },
    },
    'loggers': {
        'django.db': {
            'level': 'DEBUG',
            'handlers': ['file'],
            'propagate': True,
        },
        'netbox': {
            'level': 'DEBUG',
            'handlers': ['file'],
        },
    },
}
```

... unfortunately I didn't get any SQL logs.  Sorry, I haven't been able to fix this - maybe someone else here knows how.

**Reply by tfboy** [reply]
> > The IPs and even the "5.1" bit does not appear in any description or DNS. It's only referenced as an IP address.
> 
> Could you please double check as what you are suggesting shouldn't be possible.

My apologies, for the IP section _**does**_ have 5.1 in the description, but as it's also part of the IP, I had just looked at that and stopped checking the description. I guess it's picking out those results because of the description values and not because of the IP.

![image](https://user-images.githubusercontent.com/6972726/172572302-1258f3d0-92db-4080-bd30-9c6e155dbea0.png)

Netbox 3.2.4. My PostgresSQL is 12.11.

So, it does pick up partials for prefixes but not for IPs where it will only pick out the IP if the description has the search string.

**Reply by candlerb** [reply]
> So, it does pick up partials for prefixes

Sometimes. See #6016

**Reply by tfboy** [reply]
> 

Yes :)
But is there a way for it to pick up partials for IP addresses? If not, is this on the roadmap? I haven't searched so apologies if there's already a ticket covering that.

**Reply by candlerb** [reply]
No, and no AFAIK.
